### PR TITLE
match FermiNet

### PR DESCRIPTION
**Pull request by szbernat**

### CODE REVIEW

File: src/deepqmc/hkext.py [hunk 1]
```diff
@@ -99,3 +100,28 @@ def __call__(self, inputs):
             if i < (n_layers - 1) or not self.last_linear:
                 out = self.activation(out)
         return out + inputs if self.residual else out
+
+
+class ResidualConnection:
```
Comment (szbernat) [inline src/deepqmc/hkext.py:105]: I think this is a bit more flexible than including the option for residual connection in MLP. Also we can unify the one and two particle stream residuals like this.

File: src/deepqmc/wf/nn_wave_function/env.py [hunk 2]
```diff
@@ -28,15 +46,33 @@ def __call__(self, diffs):
...
-        for i, (z, n_shell, n_pp_shell) in enumerate(
-            zip(mol.charges, mol.n_shells, mol.n_pp_shells)
-        ):
-            for k in range(n_pp_shell, n_shell):
-                shells.append((i, z / (k + 1)))
-        return cls(shells, **kwargs)
+        return self.mo_coeff(jnp.exp(-exponent))
+
+
+class FermiNetEnvelopes(hk.Module):
```
Comment (szbernat) [inline src/deepqmc/wf/nn_wave_function/env.py:52]: I was on the fence about making this a separate class or including it as an option in our current envelopes. In the end I think there are sufficiently many differences that having separate classes results in more readable code.

File: src/deepqmc/wf/nn_wave_function/nn_wave_function.py [hunk 3]
```diff
@@ -82,21 +83,17 @@ def __init__(
...
-            n_determinants * (n_up + n_down),
-            with_bias=False,
-            w_init=lambda s, d: hk.initializers.VarianceScaling(1.0)(s, d)
-            + jnp.ones(s),
-            name='mo_coeff',
-        )
-        self.conf_coeff = hk.Linear(
-            1, with_bias=False, w_init=jnp.ones, name='conf_coeff'
+        self.envelope = envelope(hamil.mol, n_determinants)
+        self.conf_coeff = (
```
Comment (szbernat) [inline src/deepqmc/wf/nn_wave_function/nn_wave_function.py:93]: Having `conf_coeff` is not strictly necessary, since they can be included in the backflow. FermiNet doesn't have `conf_coeff`. We'll have to reconcile this with the excited states code.

File: src/deepqmc/gnn/graph.py [hunk 4]
```diff
@@ -280,13 +280,13 @@ def GraphUpdate(
...
         nodes, edges = graph
 
-        if update_edges_fn:
-            edges = update_edges_fn(edges)
-
         if update_nodes_fn:
             aggregated_edges = aggregate_edges_for_nodes_fn(nodes, edges)
             nodes = update_nodes_fn(nodes, aggregated_edges)
 
+        if update_edges_fn:
```
Comment (szbernat) [inline src/deepqmc/gnn/graph.py:287]: We first update the nodes using the current edges, and only update the edges afterwards. This only makes a difference if `deep_features` is used. Things like this are much easier to do now that the input dimensions of the MLPs are inferred, since we don't have to worry about the FermiNet and DeepErwin first layer special cases.

File: src/deepqmc/wf/nn_wave_function/env.py (hunk 2 above)
Comment (zenoone) [reply src/deepqmc/wf/nn_wave_function/env.py:52]: I would vote for keeping the two implementations as a single class. As discussed offline the `ExponentialEnvelopes` are a special case of the `FermiNetEnvelopes`, where some of the parameters (the exponential coefficients) are shared across orbitals.

File: src/deepqmc/wf/nn_wave_function/nn_wave_function.py (hunk 3 above)
Comment (zenoone) [reply src/deepqmc/wf/nn_wave_function/nn_wave_function.py:93]: I agree that having `conf_coeff` is merely a redundant parametrization, since we have removed the `conf` argument to the wave function ansatz. However, I would keep it, as it can be useful for further development and hence like your solution that allows for both. I was thinking that we could maybe simplify the API further by introducing a no op and a stop graph replacement for MLPs. This logic could be applied here aswell.

File: src/deepqmc/gnn/electron_gnn.py [hunk 5]
```diff
@@ -112,14 +113,19 @@ def __init__(
...
             subnet_factory_by_lbl.setdefault(lbl, subnet_factory)
         if deep_features:
-            self.u = {
-                typ: subnet_factory_by_lbl['u'](
-                    two_particle_stream_dim,
-                    residual=not first_layer,
-                    name=f'u{typ}',
+            self.u = (
+                subnet_factory_by_lbl['u'](
+                    two_particle_stream_dim, residual=not first_layer, name='u'
```
Comment (zenoone) [inline src/deepqmc/gnn/electron_gnn.py:118]: How does this work if the streams have different dimensions?

File: src/deepqmc/hkext.py (hunk 1 above)
Comment (szbernat) [reply src/deepqmc/hkext.py:105]: Ah indeed the `residual` `kwarg` of MLP still exists and should be removed. Also, the shared u network should not have `residual` `kwarg` passed to it as this is now handled outside the MLP. I'll make these changes. Is there anywhere else where the `residual` `kwarg` of MLP is used?

File: src/deepqmc/gnn/electron_gnn.py (hunk 5 above)
Comment (zenoone) [reply src/deepqmc/gnn/electron_gnn.py:118]: Ok, I see. So shared means that we do something along the lines of the concatenation update rule (not sharing parameters, which was my confusion). To me the [code](https://github.com/deepqmc/deepqmc/blob/5f90fd535232f0077759982c335fc79b8a869ed3/src/deepqmc/gnn/electron_gnn.py#L164) associated with this shared update looks a bit cumbersome. If you agree that this is nothing but the concatenate features + MLP (+ split) and the featurewise MLP update rules from the electron stream we should maybe try to come up with a concise unified building block. What do you think?

Comment (zenoone) [reply src/deepqmc/gnn/electron_gnn.py:118]: If not combining the updates we should at least align the nomenclature.

Comment (zenoone) [reply src/deepqmc/gnn/electron_gnn.py:118]: Ok, confusion is resolved. This actually corresponds to sharing parameters across two particle streams and the case that the two particle streams have different dimensions is not implemented. This could be refactored to be more concise and/or a comment could be added to explain the logic.

File: src/deepqmc/wf/nn_wave_function/env.py (hunk 2 above)
Comment (szbernat) [reply src/deepqmc/wf/nn_wave_function/env.py]: Resolved in https://github.com/deepqmc/deepqmc/pull/121/commits/752915f78e9021305b3644f31b985893732bb9f9

File: src/deepqmc/wf/nn_wave_function/nn_wave_function.py (hunk 3 above)
Comment (szbernat) [reply src/deepqmc/wf/nn_wave_function/nn_wave_function.py:93]: The problem is that if we don't have `conf_coeff`, the operation is not noop but `sum` (along the determinant axis). We could implement this like:
```
conf_coeff:
  _target_: jnp.sum
  _partial_: true
  keepdim: true
```
In the code we could then always write:
```
psi = self.conf_coeff(xs).squeeze()
```
What do you think? Why would we need a stop graph subnet?

Comment (zenoone) [reply src/deepqmc/wf/nn_wave_function/nn_wave_function.py:93]: Yes, that would be a simple and clean solution. Looks good to me. The stop graph subnet would be useful to remove the convolution kwarg from the GNN, but that's a different story.

Comment (szbernat) [reply src/deepqmc/wf/nn_wave_function/nn_wave_function.py]: Resolved in https://github.com/deepqmc/deepqmc/pull/121/commits/18845e56a67e3f7828c346c4ccaf25d3318e72bb

I've introduced the `SumPool` layer which is a `jnp.sum` but has the interface of a haiku module.

File: src/deepqmc/hkext.py (hunk 1 above)
Comment (zenoone) [reply src/deepqmc/hkext.py:103]: I think that was the only remaining use of the residual argument.

File: src/deepqmc/gnn/electron_gnn.py (hunk 5 above)
Comment (szbernat) [reply src/deepqmc/gnn/electron_gnn.py]: Would you mind adding this doing this refactor/adding the comment yourself? I think it might be quicker since you know better which aspects you want to improve.

File: src/deepqmc/conf/ansatz/ferminet.yaml [hunk 6]
```diff
@@ -40,7 +45,6 @@ omni_factory:
     edge_types:
```
Comment (zenoone) [inline src/deepqmc/conf/ansatz/ferminet.yaml:45]: Could we maybe merge the `edge_types` with the `update_features`, now that we have a removed the GNN base class?

Comment (zenoone) [reply src/deepqmc/conf/ansatz/ferminet.yaml:45]: Ok I see. Lets keep this for another PR.

**Review by zenoone — APPROVED** [review]
